### DISCUSSION TITLE
How do you setup a replicated memcached proxy?

### DISCUSSION THREAD

**Discussion by EnzoDotjs** [memcached/memcached#1152] [Q&A] [answered]
I am a sys admin looking to setup memcached with proxy support on my webservers which backend to AWS ElastiCache for Memcached.  Initially I thought ElastiCache would replicate the keys across all nodes in the cluster but testing proved this to be wrong.  
My ElastiCache cluster has 3 nodes and encryption in transit is turned off.  The engine is 1.6.22.
For memcached with proxy support, I built it 2 ways.  First by downloading off memcached.org, version 1.6.26 tgz, ./configure --enable-proxy ; make ; make install ; Second by cloning repo and cd ./memcached ; cd ./vendor ; ./fetch.sh ; cd .. ; ./autogen.sh ; ./configure --enable-proxy ; make ; make install ;
I cloned the memcached-proxylibs so that I can get the simple module for lua.  Taking the ./lib/simple/examples/basic-replicate.lua , I made a few changes:
Added to the top so that I do not need to export LUA_PATH:
-- Add Lua module path
local mc_proxy_module_path = "/usr/local/memcached-proxylibs/lib/simple/?.lua;/usr/local/memcached-proxylibs/lib/routelib/?.lua"
package.path = package.path .. ";" .. mc_proxy_module_path

Removed zone z2

Added each of the 3 ElastiCache nodes to z1

To start the memcached proxy I ran:
/usr/local/bin/memcached -vv -l localhost:11211 -o proxy_config=/etc/mc.cfg -u root -B ascii

As it starts the last several lines look like:
router: overriding default for	router_type
making backend for... mc.yta8ji.0001.apse2.cache.amazonaws.com:11211
making backend for... mc.yta8ji.0002.apse2.cache.amazonaws.com:11211
making backend for... mc.yta8ji.0003.apse2.cache.amazonaws.com:11211
setting up flat replicated zone. local: z1
setting up flat replicated zone. local: z1
setting up flat replicated zone. local: z1
setting up flat replicated zone. local: z1
<51 server listening (ascii)

To test I run telnet localhost 11211
set foo 0 3600 3
bar
and it gets STORED
get foo
VALUE foo 0 3
bar
END

All looks good but if I telnet to each of the backends:
Trying 10.80.13.144...
Connected to mc.yta8ji.0001.apse2.cache.amazonaws.com.
Escape character is '^]'.
get foo
END
quit
Connection closed by foreign host.
Trying 10.80.13.214...
Connected to mc.yta8ji.0002.apse2.cache.amazonaws.com.
Escape character is '^]'.
get foo
END
quit
Connection closed by foreign host.
Trying 10.80.13.187...
Connected to mc.yta8ji.0003.apse2.cache.amazonaws.com.
Escape character is '^]'.
get foo
END
quit
Connection closed by foreign host.

I can not get the key, so I suspect it is being stored locally and not being sent out to all the hosts in the zone which is what I expect.

I would like and my developers would like to be able to connect to localhost 11211 but all the keys sent out to all the backends.
What might I be doing wrong and how can I correct this?

**Comment by dormando** [comment, marked as answer]
Hey! Thanks for reaching out. I'm sorry about the trouble you ran into with the `simple` library. I've now ensured it's fully retired.

First, remove the `-B` since I think you're turning the proxy back off with that. (the proxy is implemented as a "protocol", which is a little obnoxious but it lets you use the local cache instance in advanced situations)

You will want routelib instead: https://github.com/memcached/memcached-proxylibs/tree/main/lib/routelib - and it should just work with the latest tarball of memcached (1.6.29), no need to build from git.

I've added a small example configuration for basic replication: https://github.com/memcached/memcached-proxylibs/blob/main/lib/routelib/examples/basic-replicate.lua

...and an FAQ-like file for explaining when memcached is replicated vs not (which is why elasticache doesn't offer this up front, though I think they do for some cost?): https://github.com/memcached/memcached-proxylibs/blob/main/lib/routelib/REPLICATE.md

I'll add a few more example files when I get time. Please let us know if this is helpful.

**Reply by dormando** [reply]
Edited to add a note about removing `-B`

**Reply by EnzoDotjs** [reply]
@dormando 

Thank you so much for this explanation, I now have a much better understanding of the proxy and memcached, especially the thought process behind the architecture.  I can see why replication might not be the most appropriate approach.  That being said we still want to replicate.

The `-B` option we found necessary after testing using telnet and not getting errors on set but it could be just because the `simple` library config was not correct.  Removing this now.

On the example you pushed up,  I took that added my 3 backends like so:
`pools{
    set_all = {
        { backends = { "mc.yta8ji.0001.apse2.cache.amazonaws.com:11211" } },
        { backends = { "mc.yta8ji.0002.apse2.cache.amazonaws.com:11211" } },
        { backends = { "mc.yta8ji.0003.apse2.cache.amazonaws.com:11211" } },
    }
}`

Along with, which should probably be included in the examples: 
`package.loaded["routelib"] = nil
local r = require("routelib")`

And this, which is the solution I came up with for getting the `routelib` library into the search path, although I saw alternatives in the repo:
`local mc_proxy_module_path = "/usr/local/memcached-proxylibs/lib/simple/?.lua;/usr/local/memcached-proxylibs/lib/routelib/?.lua"
package.path = package.path .. ";" .. mc_proxy_module_path`

Testing with:
`verbose(true)
debug(true)`

When starting, I ran this:
`/usr/local/bin/memcached -vv -l localhost:11211 -o proxy_config=/etc/mc.cfg -u root`

but was getting an error:
`mcp_config_pools: start
ERROR: Failed to execute mcp_config_pools: /usr/local/memcached-proxylibs/lib/routelib/routelib.lua:167: bad argument #1 to 'dofile' (string expected, got boolean)`

I tracked this down to line 701 in `memcached-proxylibs/lib/routelib/routelib.lua`:
`load_userconfig(mcp.start_arg)`

mcp.start_arg is set to false and load_userconfig expects a string but got a boolean.
I commented that line out and was able to move past it and get it to run.  With it commented out it seems to work with the same testing I did above except `stats items` does not seem to return anything even though an item is set.

Would you please correct the `routelib` library and if you can let me know why `stats items` does not work on the proxy but does work on the backends.

Thanks!

**Reply by EnzoDotjs** [reply]
Oh and yes I had to use the 1.6.29 version of memcached as on the previous version I was getting other errors after `load_userconfig`.  I took the tarball of memcached (1.6.29) and built it with `--enable-proxy`.

**Reply by dormando** [reply]
Hey,

You shouldn't need to do any of that bodging. Please see: https://github.com/memcached/memcached-proxylibs/tree/main/lib/routelib

Startup and config was simplified compared to how `simple` used to work:
```
Now, start the proxy: memcached -o proxy_config=routelib.lua,proxy_arg=example.lua
```
(though I see a typo in the README file that I'll get rid of now)

Remove the `package.loaded` stuff, and the mc_proxy_module_path stuff, instead just give the full path on the command line to both files:

`memcached -o proxy_config=/path/to/routelib.lua,proxy_arg=/path/to/your/config.lua`

where config.lua is exactly like the example file I linked earlier.. Your config.lua does _not_ need to load routelib. The proxy will take care of reloading both files during config reload, so you don't need the package.loaded bodge anymore either.

---

stats items doesn't work because the stats commands are relative to the proxy itself: the proxy is not storing any items, the servers behind it are. You can see some useful proxy specific stats with `stats proxy` and log streams like `watch proxyevents`

**Reply by dormando** [reply]
The proxy _can_ store items directly in advanced usage (like cache tiering and hot cache scenarios described in the link below) but that functionality isn't currently exposed in routelib directly yet.

Architecture examples: https://github.com/memcached/memcached/wiki/ProxyExamples (not all supported by routelib yet, but currently possible with the raw API)

**Reply by dormando** [reply]
Also apologies: I bumped routelib a few days ago to require fixes in 1.6.29. Hopefully sometime this month routelib will start bundling with the memcached tarball, so you will only have to download it if you want to modify it or need a version newer than the tarball itself.

Since we typically do tarballs monthly it should be rare for common use cases to need to grab routelib separately in the future.

good luck!

**Reply by EnzoDotjs** [reply]
AWESOME!  I am good now and sorry I missed the proper startup command.
Reverted the commented out line for `load_userconfig`, all the other junk I added in /etc/mc.cfg and changed startup command to:
`usr/local/bin/memcached -vv -l localhost:11211 -o proxy_config=/usr/local/memcached-proxylibs/lib/routelib/routelib.lua,proxy_arg=/etc/mc.cfg -u root`

And this all works!  Thanks so much for the help!
